File: frontend/src/api/modules/system.ts
import type { RequestOptions } from '../request'
import request from '../request'

export type SystemUpdateChannel = 'stable' | 'rc' | 'beta' | 'alpha' | 'exp'

export interface SystemUpdatePolicy {
  channel: SystemUpdateChannel
  availableChannels: SystemUpdateChannel[]
}

export interface SystemVersion {
  version: string
  gitSha: string
  buildTime: string
  deploymentMode: string
  deploymentModeLabel: string
  updateChannel: string
  latestVersion: string
  hasUpdate: boolean
  updateCached: boolean
  updateWarning: string | null
}

export interface SystemUpdateDetail {
  policy: SystemUpdatePolicy
  currentVersion: string
  latestVersion: string
  hasUpdate: boolean
  deploymentMode: string
  deploymentModeLabel: string
  buildType: string
  buildTypeLabel: string
  releaseUrl: string | null
  notes: string | null
  cached: boolean
  updateSupported: boolean
  unsupportedReason: string | null
  warning: string | null
}

export interface SystemUpdateAccepted {
  operationId: string
  deploymentMode: string
  message: string
  targetVersion: string
}

export interface SystemUpdateStatus {
  previousVersion: string | null
  currentVersion: string | null
  needRestart: boolean
  operation: {
    operationId: string | null
    kind: 'update' | 'rollback' | 'restart' | null
    status: 'idle' | 'running' | 'succeeded' | 'failed'
    targetVersion: string | null
    message: string | null
    error: string | null
    startedAt: string | null
    finishedAt: string | null
  }
}

export interface SystemRestartAccepted {
  message: string
  operationId: string
}

export function getSystemVersion(options: RequestOptions = {}) {
  return request<SystemVersion>({
    url: '/api/admin/system/version',
    method: 'GET',
    ...options,
  })
}

interface SystemUpdateDetailQuery {
  refresh?: boolean
  channel?: SystemUpdateChannel
}

interface SystemUpdateTarget {
  targetVersion: string
  channel: SystemUpdateChannel
}

export function getSystemUpdateDetail(data: SystemUpdateDetailQuery) {
  return request<SystemUpdateDetail>({
    url: '/api/admin/system/update/detail',
    method: 'GET',
    params: data,
  })
}

export function performSystemUpdate(data: SystemUpdateTarget, options: RequestOptions = {}) {
  return request<SystemUpdateAccepted>({
    url: '/api/admin/system/update',
    method: 'POST',
    data,
    ...options,
  })
}

export function restartSystem(options: RequestOptions = {}) {
  return request<SystemRestartAccepted>({
    url: '/api/admin/system/restart',
    method: 'POST',
    ...options,
  })
}

export function getSystemUpdateStatus(options: RequestOptions = {}) {
  return request<SystemUpdateStatus>({
    url: '/api/admin/system/update/status',
    method: 'GET',
    ...options,
  })
}

File: backend/crates/gateway-admin/tests/use_case/system.rs
use std::sync::Mutex;

use async_trait::async_trait;

use gateway_admin::{
    model::system::{
        SystemOperationAccepted, SystemOperationState, SystemOperationStatus, SystemUpdateDetail,
        SystemUpdateStatus, SystemVersion,
    },
    ports::system::{SystemOperationError, SystemOperations, SystemUpdateEventStream},
};

#[derive(Default)]
struct RecordingSystemOperations {
    target: Mutex<Option<Option<String>>>,
    channel: Mutex<Option<gateway_admin::model::system::SystemUpdateChannel>>,
}

#[async_trait]
impl SystemOperations for RecordingSystemOperations {
    async fn version(&self) -> Result<SystemVersion, SystemOperationError> {
        Ok(SystemVersion {
            version: "1.0.0".to_owned(),
            git_sha: "unknown".to_owned(),
            build_time: "unknown".to_owned(),
            deployment_mode: "source".to_owned(),
            update_channel: "stable".to_owned(),

            latest_version: "1.0.0".to_owned(),
            has_update: false,
            update_cached: false,
            update_warning: None,
        })
    }

    async fn update_detail(
        &self,
        _: bool,
        _: Option<gateway_admin::model::system::SystemUpdateChannel>,
    ) -> Result<SystemUpdateDetail, SystemOperationError> {
        Ok(SystemUpdateDetail {
            policy: gateway_admin::model::system::SystemUpdatePolicy {
                channel: gateway_admin::model::system::SystemUpdateChannel::Stable,
                available_channels: vec![gateway_admin::model::system::SystemUpdateChannel::Stable],
            },
            current_version: "1.0.0".to_owned(),
            latest_version: "1.0.0".to_owned(),
            has_update: false,
            deployment_mode: "source".to_owned(),
            build_type: "source".to_owned(),
            release_url: None,
            notes: None,
            cached: false,
            update_supported: false,
            unsupported_reason: Some("source build".to_owned()),
            warning: None,
        })
    }

    fn update_events(&self) -> SystemUpdateEventStream {
        Box::pin(futures::stream::empty())
    }

    async fn perform_update(
        &self,
        target_version: Option<String>,
        channel: Option<gateway_admin::model::system::SystemUpdateChannel>,
        _: std::sync::Arc<dyn gateway_admin::ports::system::SystemUpdatePreflight>,
    ) -> Result<SystemOperationAccepted, SystemOperationError> {
        *self.target.lock().expect("target") = Some(target_version.clone());
        *self.channel.lock().expect("channel") = channel;
        Ok(SystemOperationAccepted::Update {
            operation_id: "operation-update".to_owned(),
            deployment_mode: "source".to_owned(),
            message: "accepted".to_owned(),
            target_version: target_version.unwrap_or_else(|| "latest".to_owned()),
        })
    }

    async fn update_status(&self) -> Result<SystemUpdateStatus, SystemOperationError> {
        Ok(SystemUpdateStatus {
            need_restart: false,
            previous_version: None,
            current_version: Some("1.0.0".to_owned()),
            operation: SystemOperationState {
                operation_id: None,
                kind: None,
                status: SystemOperationStatus::Idle,
                target_version: None,
                message: None,
                error: None,
                started_at: None,
                finished_at: None,
            },
        })
    }

    async fn rollback(
        &self,
        _: std::sync::Arc<dyn gateway_admin::ports::system::SystemUpdatePreflight>,
    ) -> Result<SystemOperationAccepted, SystemOperationError> {
        Ok(SystemOperationAccepted::Rollback {
            operation_id: "operation-rollback".to_owned(),
            message: "accepted".to_owned(),
            need_restart: true,
        })
    }

    async fn restart(&self) -> Result<SystemOperationAccepted, SystemOperationError> {
        Ok(SystemOperationAccepted::Restart {
            operation_id: "operation-restart".to_owned(),
            message: "accepted".to_owned(),
        })
    }
}

#[tokio::test]
async fn system_update_should_normalize_blank_target_to_latest() {
    let operations = std::sync::Arc::new(RecordingSystemOperations::default());
    let services = super::AdminHarness::new()
        .system(operations.clone())
        .build()
        .await;
    services
        .system()
        .perform_update(Some("   ".to_owned()), None)
        .await
        .expect("perform update");

    assert_eq!(*operations.target.lock().expect("target"), Some(None));
}

#[tokio::test]
async fn system_update_should_forward_the_confirmed_channel() {
    let operations = std::sync::Arc::new(RecordingSystemOperations::default());
    let services = super::AdminHarness::new()
        .system(operations.clone())
        .build()
        .await;
    let channel = gateway_admin::model::system::SystemUpdateChannel::Beta;
    services
        .system()
        .perform_update(Some(" 1.2.0-beta.1 ".into()), Some(channel))
        .await
        .expect("accepted");
    assert_eq!(
        *operations.target.lock().expect("target"),
        Some(Some("1.2.0-beta.1".into()))
    );
    assert_eq!(*operations.channel.lock().expect("channel"), Some(channel));
}
